### ISSUE TITLE
Border is rendered with wrong color for sticky windows on empty workspace

### ISSUE THREAD

**Issue by aksr** [i3/i3#1924]
**i3 version 4.10.4-332-g7de15b3**

**Problem 1** - Example in the docs:
  *for_window [class="urxvt"] resize set 640 480*
should be changed to:
  *for_window [class="URxvt"] resize set 640 480*

**Problem 2** - *bind*/*bindcode* doesn't work:
For example:
  *bind Mod+24 sticky toggle*
or
  *bindcode Mod+24 sticky toggle*
*both* report an error.

**Problem 3** - unable to toggle (off) sticky property:
Steps to reproduce:
  1. *sticky toggle* on a window
  2. go to an empty workspace
  3. try to execute - *sticky toggle* (this time to off) - it does not work.

I'm pretty sure this isn't intended.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by Airblader** [comment]
* Problem 1: Send a pull request.
* Problem 2: "Report an error" isn't really sufficient. *What* error?
* Problem 3: Yes, this *is* intended. The command needs to be executed on the window just like any other command. Stickiness isn't a globally unique feature; any number of windows can be made sticky.

Also please open separate tickets for separate issues in the future.

**Comment by aksr** [comment]
I was being polite.
**Problem 2**:
**Problem 3**: 
If you go to an empty workspace, sticky window is focused by default.

**Comment by Airblader** [comment]
> Problem 2:

I'm not sure what to make of that. Again, what error?

> If you go to an empty workspace, sticky window is focused by default.

It shouldn't, and isn't for me. Can you please provide a log file? See the bot's comment above.

**Comment by aksr** [comment]
Then what is focused if you go to an empty workspace?

**Comment by Airblader** [comment]
> Then what is focused if you go to an empty workspace?

On an EWMH level – nothing. On an X11 level, an invisible support window (also a recent change, it used to be the root window). But for all intents and purposes it's "nothing".

**Comment by aksr** [comment]
Here, if a workspace has other windows, sticky window(s) is/are not focused - but, if it's an empty workspace, sticky window(s) is/are focused.

**Comment by Airblader** [comment]
So if you open a terminal, type in `i3-msg sticky toggle`, move your cursor out of the way and switch to an empty workspace, you can type and it will show up in the terminal – correct?

Either way, we won't get anyhwere if we don't have a log file (but please follow the instructions in the debugging docs) since I cannot reproduce either of your scenarios. And you still need to tell us what the error is that you're getting.

**Comment by aksr** [comment]
Just use whatever code for key-binding (**Problem 2**):
`bind Mod+24 sticky toggle`
24 is q

**Comment by Airblader** [comment]
It's really not our job to find out what the error is you're seeing. Why can't you just tell us what it is? Anyway, `Mod` is not a valid modifier key.

**Comment by Airblader** [comment]
Are you going to provide a log file seeing how you didn't even want to copy paste an error message? If not, please just say so.

**Comment by aksr** [comment]
I already **reported** the error:
http://sprunge.us/jAVe

**Comment by aksr** [comment]
Ok, I was wrong for the **Problem 2**. There was a typo.

**Comment by Airblader** [comment]
> I already reported the error:

No, you did not. You said "reports an error". That is not sufficient for a bug report. If you see an error, state *what* that error is. And provide a log file as the bot asks you to.

The file you linked to seems clear to me. `sum` is not a valid keysym and most certainly not a valid keycode. It also has nothing to do with `Mod+24` you stated before, but I already told you why that doesn't work anyway.

You have been asked to provide a log file for the third problem four times now. I will ask one last time: will you provide a proper log file as described in the debugging docs, yes or no? You don't need to actually provide it right away, but I would really appreciate a definite statement whether you are going to do it at all rather than ignoring this request. I cannot reproduce the issue so the only way to work on this is with a log file and your cooperation.

**Comment by aksr** [comment]
I will, just not atm.

**Comment by Airblader** [comment]
Thank you. We'll just wait on it then.

**Comment by aksr** [comment]
I think this is it:
http://logs.i3wm.org/logs/5634387206995968.bz2

**Comment by Airblader** [comment]
Yeah, thanks for separating the two issues. I've briefly looked at your log file. `sticky toggle` is only executed once in there and it sets (not unsets) stickiness on the window. Are you sure you're toggling the correct window and that it was sticky before?

If so, please repeat the procedure and use `xprop` or `xwininfo` to figure out the window id of the window in question. Then provide that window id and a new log file (sorry, this is necessary since the window ids are different everytime).

Also, just to be clear, this is what you're doing, correct?

1. Open a terminal and make it floating.
2. `sticky toggle`
3. `workspace x`
4. `sticky toggle`

Please also try to reproduce this by explicitly clicking into the sticky window before trying to toggle it to unset the stickiness, just so we can make sure that it's focused.

**Comment by Airblader** [comment]
Oh please also wait for a bit after you switched to the empty workspace, then execute the command and write down the exact time (including second) you executed the command (and tell us that time of course). Just to make it easy to see what happens right when you do that.

**Comment by aksr** [comment]
**Procedure**
  1. Open a terminal.
  2. `sticky toggle`
  3. `floating toggle`
  4. (go to) `workspace X` (it must be empty)
  5. `sticky toggle` (again)

**More info**
  * The border color is, as if the window is focused, but, input cursor rectangle of the terminal is hollow.
  * If I explicitly click on the window, and then execute `sticky toggle`, stickiness is unset.

**In short, it seems the problem is the border - its color is as if the window is focused.**

**Comment by Airblader** [comment]
I think there's no need for the new log file and so on for now. What you describe clearly shows (as you pointed out now, too) that the window doesn't have the input focus – as it should be – which is why toggling stickiness doesn't work until you focus it.

The border color thing I can reproduce. It's probably a bug in the rendering code.

**Comment by aksr** [comment]
Good. ;)

**Comment by Airblader** [comment]
> Btw, is it intended for focus mode_toggle not to work to focus sticky window in an empty workspace?

No, that's not intended. It'd be great if you opened a new window for it so we can keep them a bit separated. Sticky windows are new, I totally expected some initial hiccups. The sooner we find them, the better, since there was no release with them yet.

**Comment by Airblader** [comment]
The reason we render the container with the focused border is because in this situation, the workspace is focused in i3 and we render with the focused colorclass if the container is focused or within the focused container, the latter of which of course will be true.

That makes this tough. We don't want to change that rendering behavior because that's just what makes sense. On the other hand, we can't change what is focused in this case:
* If we focus the root container, we have the same problem.
* The EWMH support window cannot be focused as it is not a managed window.
* If we just actually focus nothing, we have a big paradigm break which requires quite a few assumptions to be changed and will likely break external tools using the IPC.

I am going to close this issue as a wontfix since there is no good way to approach this without breaking existing and useful behavior. I think the case of this happening is quite limited as you would usually switch to a new workspace to open a window there, not to see only the sticky window (but mode_toggle should of course work – different issue).

Please also note that not focusing sticky windows, even on empty workspaces, is deliberate as we would otherwise heavily annoy the user.

Looping @stapelberg in case he sees this differently, but closing for now. Sorry for that, @aksr , but thank you for taking the time to report and helping out. Please open a new issue for mode_toggle so we can look at that and get it fixed.

**Comment by Airblader** [comment]
@aksr I've reported the mode_toggle issue in #1927 myself already since I'll be providing a fix.

**Comment by aksr** [comment]
I was thinking about this...
Could the window at least be *relly* focused (on an empty workspace)?
(That way the focus border will be *the focus border*.

**Comment by Airblader** [comment]
This is what I wrote here:

> Please also note that not focusing sticky windows, even on empty workspaces, is deliberate as we would otherwise heavily annoy the user.

But I think the reasons for what would annoy the user I had in mind are actually incorrect. Maybe focusing the sticky window on an empty workspace makes sense. Let's explore this.

**Comment by aksr** [comment]
For example, *bspwm* focuses the sticky window on an empty workspace.

**Comment by stapelberg** [comment]
Isn’t focus kept when switching workspaces and the sticky window is focused? If so, then focusing sticky windows on an empty workspace seems like an anti-pattern to me. It means focus won’t be restored when you accidentally switch to an empty workspace and then back to a non-empty workspace.

**Comment by Airblader** [comment]
@stapelberg No, sticky windows don't keep the focus, which is deliberate. If the user does something in the sticky window and switches workspace, they most likely do so because they want to do something in another window on that workspace.

Something similar was my reason for not focusing sticky windows on empty workspaces, but I just noticed that my assumptions were incorrect, so I think it's worth exploring this.

**Comment by Airblader** [comment]
/edit: On the other hand, what I described above is usually a situation where the user doesn't have the sticky window focused anyway. So maybe keeping focused sticky windows focused on the destination workspace is the better solution. Let's explore that first.

**Comment by aksr** [comment]
Thinking about this further:
  1. If the sticky window is focused, going on another (non-empty) workspace - sticky window would *stay focused*.
  2. If the sticky window isn't focused, going on another (non-empty) workspace -sticky window would stay *unfocused*.
  3. If the sticky window is focused, going on another (empty) workspace - sticky window would *stay focused*.
     (This way *the focused border* stays to be *the focused border*.)
     (Opening another window, sticky window loses focus as is the standard case.)
     (This is the complement of the rule **1.**.)
  4. If the sticky window is not focused, going on another (empty) workspace - sticky window would *became focused*.
     (Something must be focused, if there is at least one window on a workspace.)
     (Sticky window would lose the focus, if you open another window (standard case).)

**Reasons for using sticky windows**
  1. Your, for example, editor, have sticky property - you want to change something (but you need a window from another workspace).
     (Sticky window is focused and it stays focused (you're editing something).)
  2. You executed `sticky toggle` on some window, then you focused another window (on the same workspace) then you jumped to some other workspace.
     (Usually you're waiting for something (some download/build progress etc.).)
     (Sticky window is not focused and doesn't have to be focused.)

This is my take on this.
I'll think about this more.

**Comment by aksr** [comment]
In short: 
  1. You use sticky window property to change something in sticky window.
  2. You use sticky window to monitor something.

* In the first case, sticky window is focused and needs to stay focused.
The first case is determined when the window it's already focused.
* In the second case - it does not have to became focused (and it usually isn't already focused).
The second case is also determined by the focus.

On an empty workspace, if you monitor something, you'll create/open another window and the sticky window would lose the focus.
If you're editing something, it's already focused, and, also - something needs to be focused.
